### PR TITLE
Filter out default configs when overrides exist.

### LINKED ISSUE #20092: PR #18772 breaks `sql_alchemy_conn_cmd` config
### Apache Airflow version

2.2.1

### What happened

#18772 added two options to `tmp_configuration_copy()` and defaults tasks run as the same user to generate the temp configuration file for the task runner with the options `include_env=False` and `include_cmds=False`. This is a change from the previous defaults which materialized the values of `*_cmd` configs into the JSON dump.

This presents a problem because when using `sql_alchemy_conn_cmd` to set the database connection and running as the same user the temporary config JSON dump now includes both `sql_alchemy_conn`, set to the Airflow distribution default as well as the user set `sql_alchemy_conn_cmd`. And because bare settings take precedence over `_cmd` versions while the Airflow worker can connect to the configured DB the task runner itself will instead use a non-existent SQLite DB causing all tasks to fail.

TLDR; The temp JSON config dump used to look something like:

```
{
  "core": {
    "sql_alchemy_conn": "mysql://...",
    ...
  }
}
```

Now it looks something like:

```

{
  "core": {
    "sql_alchemy_conn": "sqlite:////var/opt/airflow/airflow.db",
    "sql_alchemy_conn_cmd": "/etc/airflow/get-config 'core/sql_alchemy_conn'"
    ...
  }
}
```

But because `sql_alchemy_conn` is set `sql_alchemy_conn_cmd` never gets called and tasks are unable to access the Airflow DB.

### What you expected to happen

I'm not quite sure what is the preferred method to fix this issue. I guess there are a couple options:
- Remove the bare `sensitive_config_values` if either `_cmd` or `_secret` versions exist
- Ignore the Airflow default value in addition to empty values for bare `sensitive_config_values` during parsing
- Go back to materializing the sensitive configs

### How to reproduce

With either Airflow 2.2.1 or 2.2.2 configure the DB with `sql_alchemy_conn_cmd` and remove the `sql_alchemy_conn` config from your config file. Then run the Airflow worker and task runner as the same user. Try running any task and see that the task tries to access the default SQLite store instead of the configured one.

### Operating System

Debian Bullseye

### Versions of Apache Airflow Providers

_No response_

### Deployment

Virtualenv installation

### Deployment details

_No response_

### Anything else

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by xyu**
When sending configs to Airflow workers we materialize a temp config file. In #18772 a feature was added so that `_cmd` generated secrets are not written to the files in some cases instead favoring maintaining the raw `_cmd` settings. Unfortunately during materializing of the configs via `as_dict()` Airflow defaults are generated and materialized as well including defaults for the non `_cmd` versions of some settings. And due to Airflow setting precedence stating bare versions of settings winning over `_cmd` versions it results in `_cmd` settings being discarded:
https://airflow.apache.org/docs/apache-airflow/stable/howto/set-config.html

This change checks `_cmd`, env, and secrets when materializing configs via `as_dict()` so that if the bare versions of the values is exactly the same as Airflow defaults and we have "hidden" / special versions of these configs that are trying to be set we remove the bare versions so that the correct version can be used.

closes: #20092
related: #18772 #4050

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Comment by xyu** [comment]
Merged trunk in because #21664 was needed to fix broken tests.

**Review by potiuk — APPROVED** [review]
LGTM - but I think description about this behaviour should be added to "as_dict" docstring/

**Comment by potiuk** [comment]
And we need another committer chiming in.

**Comment by uranusjr** [comment]
It’d be nice if a docstring could be added to explain what `filter_by_source` does and why it is needed.

**Comment by xyu** [comment]
@potiuk / @uranusjr can you take a look at the docstrings I've added to see if they make sense? Also if one of you can trigger the full test run after the latest merge that would be awesome and let me know if there's anything else I should patch up with this PR to get it ready.

**Comment by potiuk** [comment]
After looking agian - I think what is also missing here is a bit "user" description in the https://airflow.apache.org/docs/apache-airflow/stable/howto/set-config.html 

I think we might have a lot of questions from users about the behaviour of config values as this is a "user" thing. I doubt our users will look in detail in "docstring" - they will mostly look at the documentation to understand how configuration precedence works, and especially - there should be a documentation explaining that it has changed in 2.3.0.

Would it be possible tha tyou add such "user" facing documentation @xyu

**Comment by xyu** [comment]
> there should be a documentation explaining that it has changed in 2.3.0.

I'm not quite sure what you mean @potiuk, as you pointed out in the original issue (https://github.com/apache/airflow/issues/20092#issuecomment-997431778) the precedence for configuration options is bare setting over `_cmd` or secret versions. This PR filters out the bare configs if they are set to Airflow’s built in defaults which then allows the user set `_cmd` overrides to work again. So this is really fixing a bug not adding or changing config behavior. (I mean I guess it is changing the config behavior but it's changing the behavior to an unbroken as documented state.)

**Comment by potiuk** [comment]
> I'm not quite sure what you mean @potiuk 

I mean that we should mention it was NOT working as expected before.

**Comment by xyu** [comment]
Oh I see as in for a changelog? Or do you mean add another "note" box to [this page](https://airflow.apache.org/docs/apache-airflow/stable/howto/set-config.html)?

**Comment by potiuk** [comment]
> Oh I see as in for a changelog? Or do you mean add another "note" box to [this page](https://airflow.apache.org/docs/apache-airflow/stable/howto/set-config.html)?

I think would be nice to mentiont it on this page that it was working differently before 2.3.0

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
Random environmental problems. Merging.
